tests/unit/test_read_rows_query.py

```diff
@@ -0,0 +1,284 @@
+        self.assertIn(input_batch[1], query.row_keys)
+        self.assertIn(input_batch[2].encode(), query.row_keys)
+        self.assertIn(b"test_row4", query.row_keys)
+        self.assertIn(b"test_row5", query.row_keys)
+
+    def test_add_rows_invalid(self):
+        query = self._make_one()
+        with self.assertRaises(ValueError):
+            query.add_rows(1)
+        with self.assertRaises(ValueError):
+            query.add_rows(["s", 0])
+
+    def test_duplicate_rows(self):
+        # should only hold one of each input key
+        key_1 = b"test_row"
+        key_2 = b"test_row2"
+        query = self._make_one(row_keys=[key_1, key_1, key_2])
+        self.assertEqual(len(query.row_keys), 2)
+        self.assertIn(key_1, query.row_keys)
+        self.assertIn(key_2, query.row_keys)
+        key_3 = "test_row3"
+        query.add_rows([key_3 for _ in range(10)])
+        self.assertEqual(len(query.row_keys), 3)
+
+    def test_add_range(self):
+        # test with start and end keys
+        query = self._make_one()
+        self.assertEqual(query.row_ranges, [])
+        result = query.add_range("test_row", "test_row2")
+        self.assertEqual(len(query.row_ranges), 1)
+        self.assertEqual(query.row_ranges[0][0].key, "test_row".encode())
+        self.assertEqual(query.row_ranges[0][1].key, "test_row2".encode())
+        self.assertEqual(query.row_ranges[0][0].is_inclusive, True)
+        self.assertEqual(query.row_ranges[0][1].is_inclusive, False)
+        self.assertEqual(result, query)
+        # test with start key only
+        result = query.add_range("test_row3")
+        self.assertEqual(len(query.row_ranges), 2)
+        self.assertEqual(query.row_ranges[1][0].key, "test_row3".encode())
+        self.assertEqual(query.row_ranges[1][1], None)
+        self.assertEqual(result, query)
+        # test with end key only
+        result = query.add_range(start_key=None, end_key="test_row5")
+        self.assertEqual(len(query.row_ranges), 3)
+        self.assertEqual(query.row_ranges[2][0], None)
+        self.assertEqual(query.row_ranges[2][1].key, "test_row5".encode())
+        self.assertEqual(query.row_ranges[2][1].is_inclusive, False)
+        # test with start and end keys and inclusive flags
+        result = query.add_range(b"test_row6", b"test_row7", False, True)
+        self.assertEqual(len(query.row_ranges), 4)
+        self.assertEqual(query.row_ranges[3][0].key, b"test_row6")
+        self.assertEqual(query.row_ranges[3][1].key, b"test_row7")
+        self.assertEqual(query.row_ranges[3][0].is_inclusive, False)
+        self.assertEqual(query.row_ranges[3][1].is_inclusive, True)
+        # test with nothing passed
+        result = query.add_range()
+        self.assertEqual(len(query.row_ranges), 5)
+        self.assertEqual(query.row_ranges[4][0], None)
+        self.assertEqual(query.row_ranges[4][1], None)
+        # test with inclusive flags only
+        with self.assertRaises(ValueError):
```

Comment:
`add_range` seems to be raising ValueError with varying messages. Perhaps we can also check which error message is being raised in each of the errors below.

Reply:
good point, I added that to all the raises statements